glommio/src/io/glommio_file.rs

```diff
@@ -205,6 +356,77 @@ impl GlommioFile {
         }
     }
 
+    async fn flock(&self, op: &'static str, flags: libc::c_int) -> Result<()> {
+        let guard = self
+            .lock_state
+            .as_ref()
+            .unwrap()
+            .mark_locking(self.path.borrow().as_ref().unwrap())?;
+
+        let source = self
+            .reactor
+            .upgrade()
+            .unwrap()
+            .flock(self.as_raw_fd(), flags)
+            .await;
+        source.collect_rw().await.map_err(|source| {
+            GlommioError::create_enhanced(
+                source,
+                op,
+                self.path.borrow().as_ref(),
+                Some(self.as_raw_fd()),
+            )
+        })?;
+
+        guard.commit();
+
+        Ok(())
+    }
+
+    pub(crate) async fn try_lock_shared(&self) -> Result<OwnedGlommioFile> {
+        // NOTE: The try variant could just do the syscall directly instead of dispatching to a blocking thread.
+        // For consistency though it's implemented the same was.
+        self.flock("try_lock_shared", libc::LOCK_SH | libc::LOCK_NB)
+            .await
+            .map(|()| self.clone().into())
+    }
+
+    pub(crate) async fn try_lock_exclusive(&self) -> Result<OwnedGlommioFile> {
+        // NOTE: The try variant could just do the syscall directly instead of dispatching to a blocking thread.
+        // For consistency though it's implemented the same way.
+        self.flock("try_lock_exclusive", libc::LOCK_EX | libc::LOCK_NB)
+            .await
+            .map(|()| self.clone().into())
+    }
+
+    pub(crate) async unsafe fn funlock(&self) -> Result<()> {
+        let lock_state = self.lock_state.as_ref().unwrap();
+        assert_eq!(
+            lock_state.state(),
+            AdvisoryLockState::Locked,
+            "Not holding the lock!"
+        );
+
+        let source = self
```

Comment:
One thing I've noticed here is that the background thread dispatch is quite expensive (the syscall takes microseconds but the round-trip to the background thread takes >10ms on my 3.9 GhZ CPU & the background thread should be idle). While that's fine for acquiring the lock (can argue the latency is just the cost of acquisition), it's a bit unfortunate for unlock - not sure if I should do the funlock explicitly since `try_take_last_clone_unlocking_guard ` is much slower than drop + try_take_last_clone (except that pattern isn't quite as safe if you're trying to enforce a clean shutdown).

Reply:
is the main cpu busy? My suspicion is that you have a too high poll period, and if the main thread is busy, it will not know that it is completed until poll happens.

Reply:
I am not very familiar with advisory locks in practice. Never used them, so it's hard for me to talk about the caveats. But it seems to me that:
* you have state in userspace that tracks whether the lock is held (which is great, btw)
* you will only try to acquire the lock in the kernel when you already know you will succeed.

If that's the case, maybe we can just call the blocking call? Blocking for microseconds is absolutely not a problem, especially if this is a lock (not called thousands a time per second). You'd want this to happen on the blocking thread if this is an operation that can take dozens of milliseconds, potentially.

Is there any case where we are concerning that locking (or unlocking) could take more than 1ms ?

Reply:
Benchmarking it on my machine (which isn't the best but it's the tool I have without resorting to reading the Linux kernel source), it's ~100ns. I originally had the blocking version also in this PR but that approach doesn't work because we can easily deadlock since we run out background threads to distribute work to. Probably no way to implement those without proper I/O uring support (or doing something silly like doing a try lock & then spawning a completely new thread just for waiting if it fails).

So yeah, I'll move these back onto the main thread.

Reply:
Actually, just want to confirm. While it's fine for local files, this may show problems for NFS/SMB shares of which I don't have any to really test since those will map to network I/O. Handle it via documentation or some other way?

Reply:
I could use fstatfs to determine if it's NFS/SMB & fall back to the blocking path in that case.

Reply:
I'm going to assume NFS/SMB is an unlikely use case for DmaFile for now so leaving it out. Can always add a fstatfs call later to distinguish network FS from local at the cost of an extra syscall in the advisory lock path.